src/mem/cache/Cache.py

```diff
@@ -112,7 +112,7 @@ class BaseCache(ClockedObject):
         "Notify the hardware prefetcher on every access (not just misses)",
     )
     prefetch_on_pf_hit = Param.Bool(
-        True, "Notify the hardware prefetcher on hit on prefetched lines"
+        False, "Notify the hardware prefetcher on hit on prefetched lines"
```

Comment:
I have test the `riscv-ubuntu-run.py` in the local machine and it works well by change the default value of `prefetch_on_pf_hit` to `False`. Cloud you just simply change it?

Reply:
This sounds plausible. @rjc-arch is looking into this and trying to understand what's going on here. Our current theory is that there is  bug triggered by `prefetch_on_pf_hit` related to which packets the prefetcher processes. It seems like the problem might be caused by a request received through the `ppHit` listener which the prefetcher thinks should have data but doesn't. My guess is that the hardware prefetcher doesn't filter out some cache maintenance (or similar) operation correctly.

I would suggest that we either revert the specific change that enables `prefetch_on_pf_hit` or wait for @rjc-arch to produce a patch.

Reply:
I *don't* like the idea of just changing `prefetch_on_pf_hit` to false as  then we'll have code that is broken in gem5. I would prefer to go ahead and revert this whole patch if we can't get a fix quickly.

Reply:
Hi @powerjg. Just to clarify, changing `prefetch_on_pf_hit` to False _would_ be reverting the whole commit (but preserving the other commits in the pull request, which are on the same theme but separate).

Having said that, we do suspect the parameter `prefetch_on_pf_hit = True` is exposing a separate bug or edge case in the prefetcher code, rather than being a bug itself, so we need to look into the underlying cause as well.

Reply:
Or put differently, reverting all of the changes in the PR won't remove underlying broken code, it'll just disable it by setting `prefetch_on_pf_hit` to false again.

Reply:
Hi, I have trace the `riscv-ubuntu-run.py` with `--debug-flags=Cache`, the following is the output:
```
5261728243428: board.cache_hierarchy.l1dcaches0: sendMSHRQueuePacket: MSHR WriteReq [1078fb2e0:1078fb2e7]
5261728243428: board.cache_hierarchy.l1dcaches0: createMissPacket: created UpgradeReq [1078fb2c0:1078fb2ff] from WriteReq [1078fb2e0:1078fb2e7]
5261728243428: board.cache_hierarchy.dptw_caches0: handleSnoop: snoop hit for UpgradeReq [1078fb2c0:1078fb2ff], old state is state: 6e616324 (S) writable: 0 readable: 1 dirty: 0 prefetched: 0 | tag: 0x20f1f6 secure: 0 valid: 1 | set: 0xb way: 0
5261728243428: board.cache_hierarchy.dptw_caches0: new state is state: 6e616320 (I) writable: 0 readable: 0 dirty: 0 prefetched: 0 | tag: 0xffffffffffffffff secure: 0 valid: 0 | set: 0xb way: 0
5261728243428: board.cache_hierarchy.l2caches0: access for UpgradeReq [1078fb2c0:1078fb2ff] hit state: 6 (E) writable: 1 readable: 1 dirty: 0 prefetched: 1 | tag: 0x41e3e secure: 0 valid: 1 | set: 0xcb way: 0x5
```

The gem5 crashed on the `ppHit->notify` because of the `UpgradeReq` and the cache block is *prefetched*.

Reply:
Thanks for the extra information, @rogerchang23424. I've only just managed to re-produce this problem now - it looks like the gem5 resource manager was picking up an old disk image which doesn't exhibit the problem from somewhere on my system. After purging all the copies I could find it's now failing with the assertion.

I'm running it up in a debugger now, then I'll take a proper look.